### PR TITLE
Add Data Parallel support for Torch backend

### PR DESCRIPTION
**Pull request by buildwithsuhana**
## Description
This PR adds data parallelism support when using Keras with the PyTorch backend. 

### Key additions:

- **DDP Wrapper:** In the training loop, Keras now detects active distribution strategies and wraps the model with PyTorch's native DistributedDataParallel (DDP).

- **Custom Step Routing:** Steps like train_step, test_step, and predict_step have been updated to check for this ddp_model. If found, they pass data through it, letting PyTorch handle gradient synchronization automatically.

- **Process Group Translation:** Keras DeviceMesh is translated into PyTorch Process Groups to ensure all devices coordinate perfectly.

- **Metrics Synchronization:** In get_metrics_result, metric variables are automatically reduced (summed) across all devices in the process group, ensuring evaluation logs are accurate for the whole batch.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/torch/trainer.py [hunk 1]
```diff
@@ -144,10 +202,58 @@ def test_function(data):
...
+                    results = super().get_metrics_result()
+
+                    results = {
+                        k: val.clone() if isinstance(val, torch.Tensor) else val
+                        for k, val in results.items()
+                    }
+
+                    # Restore original values.
+                    for v, original_val in zip(all_vars, original_values):
+                        v._value.copy_(original_val)
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: Hmm... I think this calls for a `StatelessScope` so that you don't have to modify the actual values of variables and then restore them.

Super rough outline:
```python
reduced_vars = [dist.all_reduce(v._value...) for v in self.metrics_variables]
with backend.StatelessScope(
            state_mapping=[
                (v, reduced_v)
                for v, reduced_v in zip(self.metrics_variables, reduced_vars)
            ]
        ) as scope:
            results = super().get_metrics_result()
```

Similar code:
https://github.com/keras-team/keras/blob/master/keras/src/backend/jax/trainer.py#L101-L113

File: keras/src/backend/torch/trainer.py [hunk 2]
```diff
@@ -38,18 +45,63 @@ def _should_torch_compile(self):
...
+                    backend_mesh = _to_backend_mesh(
+                        active_distribution.device_mesh
+                    )
+                    process_group = backend_mesh.get_group(
+                        active_distribution.batch_dim_name
+                    )
+
+                object.__setattr__(
+                    self,
+                    "ddp_model",
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py:85]: Why use `__setattr__` and `getattr` everywhere instead of just `self.ddp_model`?

File: keras/src/backend/torch/trainer.py [hunk 3]
```diff
@@ -508,13 +616,42 @@ def test_on_batch(
...
         batch_outputs = tree.map_structure(
             backend.convert_to_numpy, batch_outputs
         )
         return batch_outputs
 
 
+def _distribute_data(data):
+    return tree.map_structure(
+        lambda x: convert_to_tensor(x) if x is not None else None, data
+    )
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: I'm a bit confused about this.
- Why is it called `distribute_data` when it's not distributing. Is it coming in the next PR?
- Why was the `convert_to_tensor` not needed before? Is it because `convert_to_tensor` is used to gather a DTensor?

File: keras/src/backend/torch/trainer.py [hunk 4]
```diff
@@ -508,13 +616,42 @@ def test_on_batch(
...
+            batch = _distribute_data(batch)
+            if next_batch is None:
+                next_batch = batch
+            else:
+                current_batch = next_batch
+                next_batch = batch
+                yield current_batch
+
+        if next_batch is not None:
+            yield next_batch
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: I don't think this will do anything on Torch.

The reason why this exists on JAX is that `jax.device_put` is async. But we currently use the synchronous `to` on Torch. So you would have to modify that to make this useful.

File: keras/src/backend/torch/trainer.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Done.

File: keras/src/backend/torch/trainer.py (hunk 2 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py:85]: Regular assignment causes an infinite loop and crashes because of how PyTorch and Keras track layers. This is a safe way to bypass those errors

File: keras/src/backend/torch/trainer.py (hunk 3 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: I have removed this logic from the current PR to avoid confusion, as it wasn't necessary for the base DDP support. This will be useful in the next PR where _distribute_data will handle the actual data distribution and sharding across the mesh.

File: keras/src/backend/torch/trainer.py (hunk 4 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Removed this from the PR.

File: keras/src/backend/torch/trainer.py [hunk 5]
```diff
@@ -144,10 +202,57 @@ def test_function(data):
...
+                    process_group = backend_mesh.get_group(
+                        active_distribution.batch_dim_name
+                    )
+
+                reduced_vars = []
+                for v in self.metrics_variables:
+                    # Use a copy for reduction to avoid modifying
+                    # the original variable.
+                    val = v._value if hasattr(v, "_value") else None
+                    if val is not None:
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: - Is it really possible for `_value` to be missing?
- Is it really possible for `_value` to be `None`?
- Why not use `v.value` instead?

File: keras/src/backend/torch/trainer.py [hunk 6]
```diff
@@ -38,18 +44,64 @@ def _should_torch_compile(self):
 
         return self.jit_compile
 
+    def _initialize_ddp(self):
+        if (
+            torch.distributed.is_initialized()
+            and getattr(self, "ddp_model", self) is self
+        ):
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: Continuing on the `__setattr__` and `getattr`:
- I really don't think you need the `getattr`, please try `self.ddp_model`
- for the setter, I understand why. Decorate `_initialize_ddp` with `@tracking.no_automatic_dependency_tracking` like here, and you should be able to do `self.ddp_model = ...`

https://github.com/keras-team/keras/blob/51470cc4dac03b9d0e2f78d2455cd1d65a951106/keras/src/layers/layer.py#L346-L347

**Review by hertschuh — COMMENTED** [review]
Last tweaks I think:

File: keras/src/backend/torch/trainer.py (hunk 6 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Done

File: keras/src/backend/torch/trainer.py (hunk 5 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: I've updated the code to use v.value directly as suggested, which simplifies the logic and safely handles uninitialized variables via placeholders.

File: keras/src/backend/torch/trainer.py [hunk 7]
```diff
@@ -38,18 +46,62 @@ def _should_torch_compile(self):
...
+                    backend_mesh = _to_backend_mesh(
+                        active_distribution.device_mesh
+                    )
+                    process_group = backend_mesh.get_group(
+                        active_distribution.batch_dim_name
+                    )
+
+                object.__setattr__(
+                    self,
+                    "ddp_model",
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py:85]: Oh, but you didn't change this one to be `self.ddp_model = DistributedDataParallel(...)`.

It should work now, right?

Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py:85]: Normal assignment causes an infinite loop because PyTorch automatically registers the model as a submodule, and here the model actually contains the trainer itself. The decorator didn't help because it only stops Keras tracking, but the crash is coming from PyTorch's registration logic.

**Review by hertschuh — APPROVED** [review]
Thanks!
